Add Github action for checking PR guidelines

amirkdv opened the pull request:
## Change Summary

A new action checks PR title and body on PR open|reopen|edit.

It depends on:

- Github API for getting PR title and body
- readthedocs/commonmark.py for converting markdown to html
- beautifulsoup4 for parsing html

Implemented checks:

- No issue reference in title,
- Correct linking verbs (eg "fix") when mentioning issues in 'Related issue number',
- No incomplete task in 'Checklist'.

## Related issue number

Fixes #3500.

## Checklist

* [x] Unit tests for the changes exist
* [x] Tests pass on CI and coverage remains at 100%
* [x] Documentation reflects the changes where applicable
* [x] `changes/<pull request or issue id>-<github username>.md` file added describing change
  (see [changes/README.md](https://github.com/samuelcolvin/pydantic/blob/master/changes/README.md) for details)
* [x] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**

amirkdv commented:
please review

amirkdv commented on .github/workflows/pr-checks.yml:
```diff
@@ -0,0 +1,41 @@
+              repo: context.repo.repo,
+              body: message
+            })
+            core.setFailed("Some PR checks failed!")
```
Workflow implication: if any of the checks fail, the whole action fails. We could just post the comment and let the action succeed.

amirkdv replied:
Having done a few rounds of using this in action, I think this is probably the desired behavior.

amirkdv commented on .github/workflows/pr-check:
```diff
@@ -0,0 +1,178 @@
+        text = ' '.join(e.text for e in section)
+        if not re.search(r'#\d+', text):
+            # ok to not reference anything
+            return None
```
Workflow implication: currently it's ok to not reference any issues here. It's also ok to reference multiple issues as long as at list one of them has a proper linking verb (eg 'fixes').

amirkdv commented on .github/workflows/pr-check:
```diff
@@ -0,0 +1,178 @@
+        return 0
+
+    print('Thank you for opening a Pull Request!\n')
+    print('Before assigning it for review, please fix the following issue(s):\n')
```
Language to be confirmed.

amirkdv commented on .github/workflows/pr-checks.yml:
```diff
@@ -0,0 +1,41 @@
+
+on:
+  pull_request:
+    types: [opened, reopened, edited]
```
Workflow implications:

* We get a @github-actions comment every time the PR body/title is edited. Including every time a Checklist task is ticked off.
* Pushing new commits doesn't trigger the PR check action. This means (I think?) that a subsequent push can mask a failing PR check in the UI.

amirkdv replied:
I think the action-run-for-each-update is fine.

In addition, pushing now _does_ trigger the action because that's how we can pick up pushes for `changes/*.md`.

amirkdv reviewed (commented):
As promised in https://github.com/samuelcolvin/pydantic/issues/3500#issuecomment-1001491081, I've marked a few places with workflow implications that I'm not sure of.

samuelcolvin commented on .github/workflows/pr-check:
```diff
@@ -0,0 +1,178 @@
+
+import requests
+from github import Github
+from bs4 import BeautifulSoup
```
Do we really need to parse html with beautiful soup? Surely the body is just markdown and we can find sections via a regex like `re.search('^#+ ', ..., flags=re.M)`?

amirkdv replied:
This is probably the most important design choice here;  glad to have this thread.

There's two decisions here:
1. should we parse markdown into a document tree or just wrangle raw text?
2. assuming parsing markdown: in what format do we want to work with the document tree?

**Re 1**: we kind of need to parse markdown properly (this is what commonmark is doing) because markdown can produce complex tree structures a la DOM. For example:

~~~markdown
## Change summary

Lorem ipsum, for example:

```python
from pydantic import BaseModel

## this is a Python comment in the code block, not a markdown section heading!
class MyModel(BaseModel):
    ...
```

In the meantime, the PR template has also changed and has a new comment in it

<!--
The comment is perfectly fine until it says something like:
* [ ] this, which should not be counted as an incomplete task
-->
~~~

We _could_ wrangle raw text and it will probably get lucky and seem fine most of the time until it breaks in confusing and unpleasant ways.

----

**Re 2**: This is a secondary choice. I only chose HTML (and beautifulsoup to work with it) for ease of use. The alternative would've been some other doc format or a markdown AST. The good thing about HTML is everyone knows about it and can reason about selectors, instead of some custom tree traversal logic.

samuelcolvin commented on .github/workflows/pr-check:
```diff
@@ -0,0 +1,178 @@
+
+        verb, issue_id = match.groups()
+        if verb not in valid_issue_ref_verbs:
+            return "Issue refs should use valid linking verbs like 'fixes #{}' not '{}'.".format(
```
personal preference, but i would always use f-strings, not `format()` these days.

samuelcolvin commented on .github/workflows/pr-checks.yml:
```diff
@@ -0,0 +1,41 @@
+        if: steps.pr_checks.outcome == 'failure'
+        with:
+          github-token: ${{ secrets.GITHUB_TOKEN }}
+          script: |
```
surely best if you can do this as part of the main action - you already have the `PyGithub` dependency and `secrets.GITHUB_TOKEN`

samuelcolvin reviewed (commented):
> External action repo?

Yes, I think this should be a standalone action in a separate rpo, I might well want to use it elsewhere in the future. 

I wouldn't try to make it too customisable if that takes you a lot of time - I think it's fine for an action like this to be pretty opinionated - you might just let people switch off specific checks if required. 

Since you're doing all the work I think you should get the credit by keeping the action on your gh profile, but I'm happy to host it on mine if you'd prefer

> Are these about the `changes/*.md`? I haven't touched these yet.

Yes. This is the most important bit I'm afraid as it's very easy to miss.

samuelcolvin commented:
please update.

amirkdv commented:
> keeping the action on your gh profile

Sounds good! I'll make a separate repo on my gh profile and add you as a contributor.

amirkdv commented:
@samuelcolvin I'll pause before proceeding with the fixes for you to see the [thread] re design (to parse markdown or to wrangle text).

If you agree with the direction, I think a natural boundary for the action could be the JSON I mentioned in the issue #3500:

* input: markdown string, a PR id, or something similar
* output: break down of markdown into "sections" and produce a json like:
  ```json
  {
    "Change summary": "<p> Lorem ipsum ... </p>",
    "Checklist": "<ul> <li> [ ] to do </li> </ul>",
    "..."
  }
  ```
Then the pydantic-specific logic will be relatively simple on top of this.

[thread]: https://github.com/samuelcolvin/pydantic/pull/3584#discussion_r776894315

amirkdv commented:
@samuelcolvin I've pulled the entire action into its own [repo] (also on [marketplace] and pre-built docker image on [dockerhub] for performance). I've added you and @PrettyWood as contributors to that repo so you can make changes as you need to.

Updates from previous version on this branch:
* resolved formatting issues
* consolidated all logic in a single action, no more JS hack
* added unit tests
* added `changes/*.md` check

Here's a [dummy PR] to see it in action!

[repo]: https://github.com/amirkdv/pydantic-pr-checks
[marketplace]: https://github.com/marketplace/actions/pydantic-pr-checks
[dockerhub]: https://hub.docker.com/r/amirkdv/pydantic-pr-checks
[dummy PR]: https://github.com/amirkdv/pydantic/pull/2

----

p.s: (out of scope) maybe pydantic could have its own Github org? If/when that happens, happy to transfer the action repo there!

amirkdv commented:
please review

samuelcolvin commented:
> p.s: (out of scope) maybe pydantic could have its own Github org? If/when that happens, happy to transfer the action repo there!

Done, @amirkdv please can you transfer the repo and I'll add you to the organisation.

samuelcolvin commented:
please update.

samuelcolvin commented:
Thanks so much for your contribution @amirkdv, in the end I took some of your code and built https://github.com/pydantic/hooky
